### ISSUE TITLE
[BUG] Camera2D zoom & smoothing problem

### ISSUE THREAD

**Issue by FEDE0D** [okamstudio/godot#2074]
When the camera2D is moving and the zoom is updated, the camera movement makes a "stop" and then continues. Note this only occurs when using a smoothing factor other than 0.

I'm using dynamic changes in the camera properties using the Tween class and this is very noticeable. I don't know if this only applies to zoom only or other properties as well.

I include a test project: https://dl.dropboxusercontent.com/u/274954519/bugZoom2D.zip
Use LEFT-RIGHT to move and 1, 2, 3 to set the zoom. Notice how the speed of the camera changes, even when setting the same zoom multiple times.

**Comment by guilhermefelipecgs** [comment]
I don't think this is a bug, I changed your code a little and I think it's pretty good now. try it.
``` GDScript
extends KinematicBody2D

var zoom = Vector2(1, 1)
var old_zoom

func _ready():
	set_fixed_process(true)

func _fixed_process(delta):
	if Input.is_action_pressed("ui_left"):
		move(Vector2(-400 * delta, 0))
	if Input.is_action_pressed("ui_right"):
		move(Vector2(400 * delta, 0))

	old_zoom = zoom

	if Input.is_action_pressed("zoom_1"):
		zoom = Vector2(1, 1)
	if Input.is_action_pressed("zoom_2"):
		zoom = Vector2(1.5, 1.5)
	if Input.is_action_pressed("zoom_3"):
		zoom = Vector2(2, 2)
		
	if old_zoom != zoom:
		get_node("Camera2D").set_zoom(zoom)
```

**Comment by guilhermefelipecgs** [comment]
Maybe this only hide the bug, I dunno, but the result is very better.

**Comment by FEDE0D** [comment]
This only hides the bug.
That snippet code was just for showing the bug.
On my oroject I was using several Cameras and using Tween to change the values of the main cam to one of the others for a smothing transition. This bug hold me back from doing that.

**Comment by guilhermefelipecgs** [comment]
Strange, I was thinking that this error is because the key press was calling the set_zoom often, but I tried this way and it's working fine.
original code with bug:
``` GDScript
...
if Input.is_action_pressed("zoom_1"):
	get_node("Camera2D").set_zoom(Vector2(1, 1))
if Input.is_action_pressed("zoom_2"):
	get_node("Camera2D").set_zoom(Vector2(1.5, 1.5))
if Input.is_action_pressed("zoom_3"):
	get_node("Camera2D").set_zoom(Vector2(2, 2))
```
new code:
``` GDScript
var zoom = Vector2(1,1)
...
func _fixed_process(delta):
	...
	if Input.is_action_pressed("zoom_1"):
		zoom = Vector2(1, 1)
	if Input.is_action_pressed("zoom_2"):
		zoom = Vector2(1.5, 1.5)
	if Input.is_action_pressed("zoom_3"):
		zoom = Vector2(2, 2)
		
	get_node("Camera2D").set_zoom(zoom)
```
I don't understood what's happening.

**Comment by guilhermefelipecgs** [comment]
I expected the camera stays with the bug.
I think the problem is the _update_scroll() method.

**Comment by guilhermefelipecgs** [comment]
Well, bug confirmed for me.

**Comment by FEDE0D** [comment]
Yeap. I came to the same conclusion.
I tried to fix it with no luck.

**Comment by guilhermefelipecgs** [comment]
_update_scroll and smoothing is not synchronized.

**Comment by guilhermefelipecgs** [comment]
I know if we remove _update_scroll or do this,
``` C++
void Camera2D::set_zoom(const Vector2 &p_zoom) {

    zoom = p_zoom;
    Point2 old_smoothed_camera_pos = smoothed_camera_pos;
    _update_scroll();
    smoothed_camera_pos = old_smoothed_camera_pos;
};
```
it's works, but I dunno if I can do this.

**Comment by FEDE0D** [comment]
You can try creating a Pull Request and see if they aprove it.
I did something similar with a bug in Parallax Background and I'm still waiting to be merged.

**Comment by guilhermefelipecgs** [comment]
Do you tested this in your project?
If it's working I'll send a pr.

2015-09-03 18:57 GMT-03:00 Federico <notifications@github.com>:

**Comment by FEDE0D** [comment]
I can confirm this still happens on the 2.0 alpha build (20151030).

**Comment by akien-mga** [comment]
Did you try the proposed patch da895e8 by @guilhermefelipecgs?

**Comment by FEDE0D** [comment]
Yes. Its working right.
@guilhermefelipecgs  Can you make a pull request for this file?
